File: web/src/types/session.ts
export interface Player {
  id: number;
  name: string;
  created_at: string;
}

export interface PlayerWithCount extends Player {
  session_count: number;
  last_played: string | null;
}

export interface ScoreReaction {
  id: number;
  session_player_id: number;
  player_id: number;
  player: Player;
  reaction: string;
  created_at: string;
}

export interface SessionPlayer {
  id: number;
  player_id: number;
  player: Player;
  score_data: Record<string, unknown>;
  total_score: number | null;
  winner: boolean;
  reactions: ScoreReaction[];
}

export interface GameBrief {
  id: number;
  name: string;
}

export interface ExpansionBrief {
  id: number;
  name: string;
}

export interface SessionImage {
  id: number;
  session_id: number;
  player_id: number | null;
  player: Player | null;
  filename: string;
  original_filename: string;
  content_type: string;
  image_url: string;
  created_at: string;
}

export interface GameSession {
  id: number;
  game_id: number;
  game: GameBrief;
  played_at: string;
  notes: string | null;
  sealed: boolean;
  sealed_at: string | null;
  created_at: string;
  players: SessionPlayer[];
  expansions: ExpansionBrief[];
  images: SessionImage[];
}

export interface SessionPlayerCreate {
  player_id: number;
  score_data: Record<string, unknown>;
}

export interface GameSessionCreate {
  game_id: number;
  played_at?: string;
  notes?: string;
  expansion_ids?: number[];
  players: SessionPlayerCreate[];
}

export interface GameSessionUpdate {
  played_at?: string;
  notes?: string | null;
  expansion_ids?: number[];
  players: SessionPlayerCreate[];
}

File: app/app/schemas/session.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel


class PlayerCreate(BaseModel):
    name: str


class PlayerUpdate(BaseModel):
    name: str


class PlayerRead(BaseModel):
    id: int
    name: str
    created_at: datetime

    model_config = {"from_attributes": True}


class PlayerReadWithCount(BaseModel):
    id: int
    name: str
    created_at: datetime
    session_count: int
    last_played: datetime | None = None

    model_config = {"from_attributes": True}


class SessionPlayerCreate(BaseModel):
    player_id: int
    score_data: dict[str, Any] = {}


class ScoreReactionRead(BaseModel):
    id: int
    session_player_id: int
    player_id: int
    player: PlayerRead
    reaction: str
    created_at: datetime

    model_config = {"from_attributes": True}


class SessionPlayerRead(BaseModel):
    id: int
    player_id: int
    player: PlayerRead
    score_data: dict[str, Any]
    total_score: int | None
    winner: bool
    reactions: list[ScoreReactionRead] = []

    model_config = {"from_attributes": True}


class GameSessionCreate(BaseModel):
    game_id: int
    played_at: datetime | None = None
    notes: str | None = None
    expansion_ids: list[int] = []
    players: list[SessionPlayerCreate] = []


class GameSessionUpdate(BaseModel):
    played_at: datetime | None = None
    notes: str | None = None
    expansion_ids: list[int] = []
    players: list[SessionPlayerCreate] = []


class SessionImageRead(BaseModel):
    id: int
    session_id: int
    player_id: int | None
    player: PlayerRead | None = None
    filename: str
    original_filename: str
    content_type: str
    image_url: str = ""
    created_at: datetime

    model_config = {"from_attributes": True}


class GameSessionRead(BaseModel):
    id: int
    game_id: int
    game: GameBrief
    played_at: datetime
    notes: str | None
    sealed: bool = False
    sealed_at: datetime | None = None
    created_at: datetime
    players: list[SessionPlayerRead]
    expansions: list[ExpansionBrief] = []
    images: list[SessionImageRead] = []

    model_config = {"from_attributes": True}


class ReactionSet(BaseModel):
    reaction: str


VALID_REACTIONS = {"🏆", "👏", "🔥", "😱", "💀", "😂", "👍", "👎", "😮"}


class GameBrief(BaseModel):
    id: int
    name: str

    model_config = {"from_attributes": True}


class ExpansionBrief(BaseModel):
    id: int
    name: str

    model_config = {"from_attributes": True}


# Rebuild models that use forward references
GameSessionRead.model_rebuild()
